ss_admin_console/main/default/classes/feeder/UnitOfWorkHelper.cls

```diff
@@ -0,0 +1,123 @@
+        Map<Integer, SObject> sampleDataMap = generateSampleRecordsMap(data, parentSObjectType);
+        for (Integer item : sampleDataMap.keySet()) {
+            SObject sampleSObject = sampleDataMap.get(item);
+            if (existentMap.containsKey(item)) {
+                SObject existentSObject = existentMap.get(item);
+                if (existentSObject != null) { // parent record exists
+                    sampleSObject.putSObject(parentSObjectType, existentSObject);
+                    if (sampleSObject.Id == null) { 
+                        uow.registerNew(sampleSObject); 
+                    } else { 
+                        uow.registerDirty(sampleSObject); 
+                    } 
+                    uow.registerRelationship(sampleSObject, supplier.getMetadata().referenceFields[0].referenceField, sampleSObject.getSObject(parentSObjectType));   
+                } 
+            } 
+        }
+        return uow;
+    }
+
+    @TestVisible
+    private Map<Integer, SObject> generateRecordsMap(List<SObject> records) {
+        Map<Integer, SObject> recordsMap = new Map<Integer, SObject>();
+        for (SObject existSObj : records) {          
+            recordsMap.put(new RecordKey(supplier.getMetadata().referenceFields[0].referencedKeyFields, existSObj).hashCode(), existSObj);
+        }
+        return recordsMap;
+    }
+
+    @TestVisible
+    private Map<Integer, SObject> generateSampleRecordsMap(List<SObject> records, String parentSObjectType) {
+        Map<Integer, SObject> recordsMap = new Map<Integer, SObject>();
+        for (SObject existSObj : records) {          
+            recordsMap.put(new RecordKey(supplier.getMetadata().referenceFields[0].referencedKeyFields, existSObj.getSObject(parentSObjectType)).hashCode(), existSObj);
+        }
+        return recordsMap;
+    }
+    @TestVisible
+    private List<SObject> getRecords(List<SObject> data) {
+        List<Set<Object>> referenceFieldValues = this.getReferenceFieldValues(data);
+        Integer fieldsNumber = referenceFieldValues.size();
+        DataDefinition.ReferenceFieldDefinition referenceFieldsList = supplier.getMetadata().referenceFields[0];
+        Object pr0 = (fieldsNumber > 0) ? SetUtils.toSetOfType(referenceFieldsList.referencedKeyFields[0], referenceFieldValues[0]) : null;
+        Object pr1 = (fieldsNumber > 1) ? SetUtils.toSetOfType(referenceFieldsList.referencedKeyFields[1], referenceFieldValues[1]) : null;
+        Object pr2 = (fieldsNumber > 2) ? SetUtils.toSetOfType(referenceFieldsList.referencedKeyFields[2], referenceFieldValues[2]) : null;
+        String query = this.buildQuery();
+        return Database.query(query);
+    }
+
+    @TestVisible
+    private List<Set<Object>> getReferenceFieldValues(List<SObject> data) {
+        List<Set<Object>> constantValues = new List<Set<Object>>();
+        SObject parentSobj;
+        for (FieldReference field : supplier.getMetadata().referenceFields[0].referencedKeyFields) {
```

Comment:
You can extract supplier.getMetadata().referenceFields[0] to local variable to make code cleaner. It's duplicated